### PR TITLE
012 add role to existing

### PR DESCRIPTION
**Pull request by sebastianczajkowski**
As an Admin
I want to add a new role to an existing capability and band
So that I can keep the software information up to date

Acceptance criteria

- Able to add a new role to an existing band and capability
- Able to define band for new role
- Able to define capability for new role
- Able to add details to new role: 
 - Job role name
 - Job spec summary

### CODE REVIEW

File: src/main/java/com/kainos/ea/backend/controllers/JobRoleController.java [hunk 1]
```diff
@@ -25,5 +27,13 @@ public JobRoleController(JobRolesService jobRolesService) {
         return jobRolesService.getAllJobRolesSortedByCapability();
     }
 
-
+    @PostMapping("/add")
+    public ResponseEntity<?> addJobRole(@RequestBody JobRole jobRole){
```
Comment (TheoHodges) [inline src/main/java/com/kainos/ea/backend/controllers/JobRoleController.java:31]: Should the <?> be there or should it read ResponseEntity<Object>?

File: src/main/java/com/kainos/ea/backend/models/JobRole.java [hunk 2]
```diff
@@ -21,10 +22,10 @@
     private String specification;
 
     @Column(name = "sharepointURL")
-    private String sharepointUrl;
+    private String sharepointUrl = "https://kainossoftwareltd.sharepoint.com/people/SitePages/Career-Lattice.aspx";
```
Comment (TheoHodges) [inline src/main/java/com/kainos/ea/backend/models/JobRole.java:25]: Would it not make more sense for it to default to a null?

File: src/main/java/com/kainos/ea/backend/controllers/JobRoleController.java (hunk 1 above)
Comment (sebastianczajkowski) [reply src/main/java/com/kainos/ea/backend/controllers/JobRoleController.java:31]: The wildcard needed to be there with the entry solution I've had, but now it is redundant so I will patch this up quick

File: src/main/java/com/kainos/ea/backend/services/JobRolesService.java [hunk 3]
```diff
@@ -1,19 +1,63 @@
...
+    public void addJobRole(JobRole jobRole) throws Exception {
+        Pattern pattern = Pattern.compile("^[A-z][A-z ]+$", Pattern.CASE_INSENSITIVE);
+        Matcher matcher = pattern.matcher(jobRole.getName());
+        if (jobRole.getName().length() > 32 || !matcher.find()) {
+            throw new Exception("Invalid role name!");
+        }
+        pattern = Pattern.compile("^[A-z][0-9A-z '().,/-]{0,249}$", Pattern.CASE_INSENSITIVE);
+        matcher = pattern.matcher(jobRole.getName());
+        if (jobRole.getSpecification().length() > 250 || !matcher.find()) {
+            throw new Exception("Invalid specification!");
```
Comment (TheoHodges) [inline src/main/java/com/kainos/ea/backend/services/JobRolesService.java:46]: Could Exception be too generic? I have used InvalidNameException for a similar thing. Either way, I think we should be consistent

File: src/main/java/com/kainos/ea/backend/models/JobRole.java (hunk 2 above)
Comment (sebastianczajkowski) [reply src/main/java/com/kainos/ea/backend/models/JobRole.java:25]: We have discussed for the default URL to be a sharepoint URL, this is why I've set it like that, we can change it after another discussion if needed

File: src/main/java/com/kainos/ea/backend/services/JobRolesService.java (hunk 3 above)
Comment (sebastianczajkowski) [reply src/main/java/com/kainos/ea/backend/services/JobRolesService.java:46]: Yup, my bad, Michał warned me as well but I forgot to change it. I'll get on it tomorrow and google through the Exceptions to see what would be the most appropriate thing here.

File: src/main/java/com/kainos/ea/backend/models/JobRole.java (hunk 2 above)
Comment (sebastianczajkowski) [reply src/main/java/com/kainos/ea/backend/models/JobRole.java:25]: After a night's worth of sleep I agree with you that I should set it to NULL and handle the nulls in the browser

**Review by TheoHodges — APPROVED** [review]

File: src/main/java/com/kainos/ea/backend/services/JobRolesService.java [hunk 4]
```diff
@@ -1,19 +1,63 @@
...
+        return jobRolesRepository.save(jobRole);
+    }
+
+    public void addJobRole(JobRole jobRole) throws IllegalArgumentException {
+        Pattern pattern = Pattern.compile("^[A-z][A-z ]+$", Pattern.CASE_INSENSITIVE);
+        Matcher matcher = pattern.matcher(jobRole.getName());
+        if (jobRole.getName().length() > 32 || !matcher.find()) {
+            throw new IllegalArgumentException("Invalid role name!");
+        }
+        pattern = Pattern.compile("^[A-z][0-9A-z '().,/-]{0,249}$", Pattern.CASE_INSENSITIVE);
```
[the remark this answers is not in the document]
Comment (sebastianczajkowski) [reply src/main/java/com/kainos/ea/backend/services/JobRolesService.java:43]: Will do that, thanks

**Review by michaltangri — APPROVED** [review]

**Pull request merged**